packages/mobile/src/identity/feelessVerification.ts

```diff
@@ -0,0 +1,1061 @@
+      [komenciKit, komenciKit.requestAttestations],
+      mtwAddress,
+      phoneHash,
+      numAttestationsRequestsNeeded
+    )
+    if (!requestTxResult.ok) {
+      Logger.debug(TAG, '@feelessRequestAttestations', 'Failed request tx')
+      throw requestTxResult.error
+    }
+
+    ValoraAnalytics.track(VerificationEvents.verification_request_attestation_request_tx_sent)
+  }
+
+  Logger.debug(`${TAG}@feelessRequestAttestations`, 'Waiting for block to select issuers')
+  ValoraAnalytics.track(VerificationEvents.verification_request_attestation_await_issuer_selection)
+
+  yield call(
+    [attestationsWrapper, attestationsWrapper.waitForSelectingIssuers],
+    phoneHash,
+    mtwAddress
+  )
+
+  Logger.debug(`${TAG}@feelessRequestAttestations`, 'Selecting issuers')
+  ValoraAnalytics.track(VerificationEvents.verification_request_attestation_select_issuer)
+
+  const selectIssuersTxResult: Result<TransactionReceipt, FetchError | TxError> = yield call(
+    [komenciKit, komenciKit.selectIssuers],
+    mtwAddress,
+    phoneHash
+  )
+
+  if (!selectIssuersTxResult.ok) {
+    Logger.debug(TAG, '@feelessRequestAttestations', 'Failed selectIssuers tx')
+    throw selectIssuersTxResult.error
+  }
+
+  ValoraAnalytics.track(VerificationEvents.verification_request_attestation_issuer_tx_sent)
+}
+
+function* feelessGetCodeForIssuer(issuer: string) {
+  const existingCodes: AttestationCode[] = yield select(feelessAttestationCodesSelector)
+  return existingCodes.find((c) => c.issuer === issuer)
+}
+
+export function* feelessCompleteAttestation(
+  komenciKit: KomenciKit,
+  mtwAddress: string,
+  phoneHashDetails: PhoneNumberHashDetails,
+  attestation: ActionableAttestation
+) {
+  const issuer = attestation.issuer
+  ValoraAnalytics.track(VerificationEvents.verification_reveal_attestation_await_code_start, {
+    issuer,
+  })
+  const code: AttestationCode = yield call(feelessWaitForAttestationCode, issuer)
+
+  ValoraAnalytics.track(VerificationEvents.verification_reveal_attestation_await_code_complete, {
+    issuer,
+  })
+
+  Logger.debug(TAG + '@feelessCompleteAttestation', `Completing code for issuer: ${code.issuer}`)
+
+  const completeTxResult: Result<TransactionReceipt, FetchError | TxError> = yield call(
+    [komenciKit, komenciKit.completeAttestation],
+    mtwAddress,
+    phoneHashDetails.phoneHash,
+    code.issuer,
+    code.code
+  )
+
+  if (!completeTxResult.ok) {
+    Logger.debug(TAG, '@feelessCompleteAttestation', 'Failed complete tx')
+    throw completeTxResult.error
+  }
+
+  ValoraAnalytics.track(VerificationEvents.verification_reveal_attestation_complete, { issuer })
+
+  yield put(feelessCompleteAttestationCode(code))
+  Logger.debug(TAG + '@feelessCompleteAttestation', `Attestation for issuer ${issuer} completed`)
+}
+
+// Get the code from the store if it's already there, otherwise wait for it
+function* feelessWaitForAttestationCode(issuer: string) {
```

Comment:
It seems odd to me that this one is separate for feeless. Could this simply take the `code` as an input instead of from the state and share the same function and acton with the other version of this?

Reply:
That's probably true. Given our desire to better integrate feeless and non-feeless verification shortly, I chose to prioritize a clear separation of state & actions over code maintainability and dryness. In places where I thought it was safe to reuse code, I went for it. When I was unsure, I defaulted to the "safe" choice of duplicating because I believe testing the use case of switching between feeless and non-feeless will be difficult.